### ISSUE TITLE
ParseUser.getCurrentUser() become null once in a while

### ISSUE THREAD

**Issue by jjalan** [ParsePlatform/Parse-SDK-Android#502]
**Problem:**
Once the user has successfully logged in to the app, once in a while (no definite pattern has been identified as to when this happens), ParseUser.getCurrentUser() becomes null even though logout is not called.

**Details**
We are using Parse SDK on Android to let user log in. This is how we initialize Parse on Application onCreate:

```
Parse.Configuration.Builder parseConfiguration = new Parse.Configuration.Builder(this)
                .applicationId(getString(R.string.parse_app_id))
                .clientKey(getString(R.string.parse_client_key))
                .server(getString(R.string.parse_server_url))
                .enableLocalDataStore();

Parse.initialize(parseConfiguration.build());`
```
In the same method, we also fetch latest info for the user:
```
if (ParseUser.getCurrentUser() != null) {
    ParseUser.getCurrentUser().fetchInBackground();
}
```

Then in our main activity, if the currentUser is null, we show login screen as follows:

```
if ( ParseUser.getCurrentUser() == null) {
    Intent intent = new Intent(this, LoginActivity.class);
    startActivityForResult(intent, LOGIN_REQUEST);
}
```

In LoginActivity, we get Facebook & Google auth tokens directly (without using Parse SDK). Once we get the facebook/google access token, we then call

```
Task<ParseUser> parseUserTask = ParseUser.logInWithInBackground("facebook", authData);
```

Once this task is completed successfully, we let user sign in and use the app. The whole flow works great. 

However, once in a while (randomly), user is shown login screen. In our app, we do not call logout or invoke the login screen workflow anywhere. Can someone please suggest what could be the possible cause for this?

If its of any help, we are hosting our server on AWS and initialize it as follows:

```
var api = new ParseServer({
  databaseURI: databaseUri
  cloud: process.env.CLOUD_CODE_MAIN || __dirname + '/cloud/main.js',
  appId: process.env.APP_ID,
  masterKey: process.env.MASTER_KEY,
  serverURL: process.env.SERVER_URL,
  push: {
	  android: {
		  senderId: process.env.GOOGLE_GCM_SENDER_ID,
		  apiKey: process.env.GOOGLE_GCM_API_KEY
	  }
  },
  filesAdapter: new S3Adapter(
	  process.env.S3_AWS_ACCESS_KEY_ID,
	  process.env.S3_AWS_SECRET_ACCESS_KEY,
	  process.env.S3_BUCKET_NAME,
	  {
		  directAccess: true, 
		  bucketPrefix: process.env.S3_BUCKET_NAME
	  }
    )
});
```

Thanks,
Jaikishan

**Comment by jjalan** [comment]
I further tried to debug it. So, if local store is enabled, Parse stores the current user details as a pin in database as shown below:

```
public ParseCurrentUserController getCurrentUserController() {
    if (currentUserController.get() == null) {
      File file = new File(Parse.getParseDir(), FILENAME_CURRENT_USER);
      FileObjectStore<ParseUser> fileStore =
          new FileObjectStore<>(ParseUser.class, file, ParseUserCurrentCoder.get());
      ParseObjectStore<ParseUser> store = Parse.isLocalDatastoreEnabled()
          ? new OfflineObjectStore<>(ParseUser.class, PIN_CURRENT_USER, fileStore)
          : fileStore;
      ParseCurrentUserController controller = new CachedCurrentUserController(store);
      currentUserController.compareAndSet(null, controller);
    }
    return currentUserController.get();
  }
```

This means that at some point in time, someone is unpinning this object. We never call logout or unpin on the ParseUser object directly. Where should we look for to debug this issue? Please help.

**Comment by btanev** [comment]
I'm having exactly the same problem. On a random pattern `ParseUser.getCurrentUser()` returns `null` without calling the logout method or without the parse session been invalidated.

**Comment by jjalan** [comment]
@btanev after spending quite a bit on this - I decided to store session token on my own locally. I also then wrapped ParseUser.getCurrentUser() in a util function - if this is null, then I would call Parse.become with that session token (if its available). If both fails, I consider user logged out.

**Comment by btanev** [comment]
Hi @jjalan

That's a nice idea I might try it! 

I'm still trying to debug the source of the issue. Right now i'm thinking what might be the reason for the random "unpinning" of the `ParseUser` (pin : `_currentUser`). I guess that the problem occurs when the `ParseUser` is extended to have additional `ParseObject`s as fields. They are not saved/pinned on the local storage along with the `ParseUser`.

Also i've found this suspicious  comment in the code while debugging the problem: https://github.com/ParsePlatform/Parse-SDK-Android/blob/master/Parse/src/main/java/com/parse/ParseObject.java#L3598

This is just a wild guess based on my instinct. I might be wrong.

Your idea is a nice workaround and if everything fails I'll use it for sure. The only problem I see with it is that when the device is "offline" and this problem occurs we are going to leave the user unable to use the application because `ParseUser.become` makes an API request.

**Comment by alefq** [comment]
Same problem found with latest version 1.15.8, released in Jun 2017. 
Doing some digging, I've found that the null pointer exception occurs when:
1) Device enters in "doze" mode, API Level 23+
2) Battery saver mode is enabled
The null in my case is caused on this line:

https://github.com/parse-community/Parse-SDK-Android/blob/aaf740aba216836a2621ed6987bf5eed09dec442/Parse/src/main/java/com/parse/Parse.java#L493

which leads to

https://github.com/parse-community/Parse-SDK-Android/blob/aaf740aba216836a2621ed6987bf5eed09dec442/Parse/src/main/java/com/parse/ParsePlugins.java#L48

`instance`  becomes null somehow.

I will try @jjalan suggestion, but I think, doing a lazy intialization on  `ParsePlugins.instance` might solve the problem.

**Comment by natario1** [comment]
@alefq Are you sure 

- you are initializing parse in the Application class `onCreate()`
- you are doing your SDK calls after parse is initialized?

**Comment by alefq** [comment]
Yes @natario1 thanks for your quick response, all of that is made on app start.
The app has several services scheduled that runs in background approximately 4 times a day, that's where `instance` becomes null, and app crashes. 
Working hours usage has no problems, most of the times the crash happens first run in the morning (after long inactivity period)  or if battery saver was activated (for a long period over one hour).

**Comment by natario1** [comment]
I would expect `Application.onCreate()` to be called when a Service is created, even in battery saver mode. And if `Parse.initialize` is there I can't see how instance could become null...

If you have other info, feel free to share. This seems related, since `ParseUser.getCurrentUser()` also calls  `ParsePlugins.get()` along the way.

**Comment by alefq** [comment]
`Application.onCreate()` is called, I initialize parse with these code, only if there's a network connection:

```
Parse.initialize(new Parse.Configuration.Builder(context)
                    .applicationId(PARSE_APPLICATION_ID)
                    .clientKey(CLIENT_KEY)
                    .server(getString(R.string.parse_server_url))
                    .enableLocalDataStore()
                    .build()
            );
```
I had already problems with session in the past, I solved most of them checking
```
    public static boolean isParseSessionActive() {
        return ParseUser.getCurrentUser() != null && ParseUser.getCurrentUser().isAuthenticated();
    }
```
Could it be, that returning from doze or in battery saver mode, there's no network instantly and Parse does not get initialized?
What are the conditions under with `instance` could become `null` within ParsePlugins lifecycle?

**Comment by natario1** [comment]
It might very well be, you should always initialize during `onCreate` with no conditions. It will do no harm.

There is no lifecycle, `instance` should be null only from process creation to the time you call `Parse.initialize()`.

**Comment by alefq** [comment]
I will make that change, thank you for the suggestion. 
I'm not quite sure that App actually stops though since other stuff gets done in background which are not Parse dependant; I'll get back to you with more comments to see how it goes.
There's must be something in newer Android API Levels' lifecycle that makes Parse's initialization goes away.

In order to prevent at least a crash, I've seen that there's this method

https://github.com/parse-community/Parse-SDK-Android/blob/aaf740aba216836a2621ed6987bf5eed09dec442/Parse/src/main/java/com/parse/Parse.java#L454

which could prevent the `NullPointerException`

On the other hand, this method is marked as `@Deprecated`

https://github.com/parse-community/Parse-SDK-Android/blob/aaf740aba216836a2621ed6987bf5eed09dec442/Parse/src/main/java/com/parse/Parse.java#L492

but newer mentioned in javadoc, will crash the same

https://github.com/parse-community/Parse-SDK-Android/blob/aaf740aba216836a2621ed6987bf5eed09dec442/Parse/src/main/java/com/parse/Parse.java#L510

**Comment by natario1** [comment]
Well dirs are not the only things that would break if parse is not initialized.
But we assume that developer initializes the SDK during `Application.onCreate()` and that method is called before anything else.

**Comment by littleariel** [comment]
Seeing the same issue with getCurrentUser() starting to return null after some time, and I was not able to pinpoint the triggering event. In my case, once getCurrentUser() starts returning null, it always does after every app restart. The only way I found to resolve the issue is to clear the app cache entirely, or uninstalling/re-installing the app.

When this bug happens, we also see errors on calls to unpinAll:
`java.lang.IllegalStateException: Attempted to find non-existent uuid d5b6348b-a0bd-425e-b950-65108b57e286`

The only viable solution I found so far is the one suggested by @jjala which is to store the session token and use become() if Parse returns null erroneously on getCurrentUser().

**Comment by qandeelabbassi** [comment]
I started facing this issue after enabling local datastore. In my case the user is not null but anonymous because I have `ParseUser.enableAutomaticUser();` in my application class.
Any update on this issue?
